### PR TITLE
2163: Add API to Repository for working "git notes"

### PR DESCRIPTION
**Pull request by edvbld**
Hi all,

please review this patch that adds the methods `addNote`. `notes` and `pushNotes` to `Repository.java` to allow Skara code to work with [git notes](https://git-scm.com/docs/git-notes). This will be used in another patch to implement a "git note"-notifier (see https://bugs.openjdk.org/browse/SKARA-2164). 

I opted for a fairly minimal API here even though the `git notes` command exposes many more features. I also decided to add the `pushNotes` method although callers could use the `push(String refspec, URI uri)` variant because I don't like the notes refspec to "leak" out into other parts of the code.

Also added a couple of unit tests.

Thanks,
Erik

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2163](https://bugs.openjdk.org/browse/SKARA-2163): Add API to Repository for working "git notes" (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [b6aabbda](https://git.openjdk.org/skara/pull/1607/files/b6aabbda38033960fe37a6c2421bd74cdd36e863)
 * [Zhao Song](https://openjdk.org/census#zsong) (@zhaosongzs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1607/head:pull/1607` \
`$ git checkout pull/1607`

Update a local copy of the PR: \
`$ git checkout pull/1607` \
`$ git pull https://git.openjdk.org/skara.git pull/1607/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1607`

View PR using the GUI difftool: \
`$ git pr show -t 1607`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1607.diff">https://git.openjdk.org/skara/pull/1607.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1607#issuecomment-1914812453)

### CODE REVIEW

File: vcs/src/main/java/org/openjdk/skara/vcs/Repository.java [hunk 1]
```diff
@@ -191,6 +191,18 @@ default void updateSubmodule(Submodule s) throws IOException {
         updateSubmodule(s.path());
     }
 
+    void addNote(Hash hash,
+                 List<String> note,
```
Comment (zhaosongzs) [inline vcs/src/main/java/org/openjdk/skara/vcs/Repository.java:195]: Just out of curiosity, is there a specific reason for using `List<String>` in the interface and then concatenating the list? Why not use `String` directly?

**Review by zhaosongzs — APPROVED** [review]
Looks good!

File: vcs/src/main/java/org/openjdk/skara/vcs/Repository.java (hunk 1 above)
Comment (edvbld) [reply vcs/src/main/java/org/openjdk/skara/vcs/Repository.java:195]: I just think it is more clear for a user of the API how lines are handled. If the method accepts `String hash` then it is (to me) harder to know if the API supports a single line or multiple lines (and the caller is responsible for inserting `\n` characters). With `List<String>` it is to me more clear, although I could have named the parameter `lines` to make it obvious 😄

**Review by zhaosongzs — APPROVED** [review]

**Comment by edvbld** [comment]
Thanks @zhaosongzs and @erikj79 for reviewing!

/integrate

**Pull request closed without merging**